Enable add/remove pairs & currencies from grpc

sangaman opened the pull request:
Closes #313. Addresses part of #424 - renaming getpairs` to `listpairs`.

This PR adds grpc calls for `addCurrency`, `addPair`, `removeCurrency`, and `removePair`. This allows the user to modify supported currencies and trading pairs without needing to edit the database manually.

It removes the `SwapProtocol` enum and instead tracks the payment channel network client used for swaps on a per-currency basis using a new `NetworkClient` enum. This is a breaking database change that will require the `currencies` and `pairs` tables to be recreated.

New service tests are added to test the various use cases for the newly added calls, along with corresponding cli commands.

michael1011 commented on lib/grpc/GrpcService.ts:
```diff
@@ -46,16 +46,20 @@ class GrpcService {
       case orderErrorCodes.DUPLICATE_ORDER:
       case p2pErrorCodes.NODE_ALREADY_CONNECTED:
+      case orderErrorCodes.CURRENCY_ALREADY_SUPPORTED:
+      case orderErrorCodes.PAIR_ALREADY_SUPPORTED:
```
I would rename those two errors to `CURRENCY_ALREADY_EXISTS` and `PAIR_ALREADY_EXISTS` because `ALREADY_SUPPORTED` doesn't sound like an error.

michael1011 commented on lib/orderbook/OrderBook.ts:
```diff
@@ -86,14 +94,21 @@ class OrderBook extends EventEmitter {
       this.peerOrders.set(pair.id, this.initOrders());
+      this.pairs.set(pair.id, pair);
+    });
+    currencies.forEach((currency) => {
```
It doesn't matter but shouldn't the currencies be initialized before the pairs?

michael1011 commented on lib/orderbook/OrderBook.ts:
```diff
@@ -136,6 +151,56 @@ class OrderBook extends EventEmitter {
+        }
+      }
+      this.currencies.delete(currencyId);
+      return currency.destroy();
```
We should declare functions that return a promise as `async`.

sangaman replied:
Stylistically I don't think there's a problem with non-async methods returning promises, I believe there are instances of that throughout the codebase and as far as I know it's not bad practice or anything and doesn't affect the compiler or type checking. Correct me if I'm wrong.

In this case specifically it's actually returning a `Bluebird` and not a promise, they function pretty much the same but aren't exactly the same I believe, and making the method `async` gives it a return type of `Promise<void>` instead of `Bluebird<void>` which seems undesirable to me.

michael1011 replied:
We can leave it as is then. I wasn't aware that not declaring `async` doesn't change its behavior.

michael1011 commented on lib/service/Service.ts:
```diff
@@ -55,9 +55,13 @@ const argChecks = {
-    const supportedCurrencies = ['BTC', 'LTC']; // TODO: detect supported currencies automatically instead of hardcoding
-    if (!supportedCurrencies.includes(currency)) throw errors.INVALID_ARGUMENT(`currency ${currency} is not supported`);
+  VALID_CURRENCY: ({ currency }: { currency: string }) => {
+    if (currency.length !== 3 || !currency.match(/^[A-Z]+$/))  {
```
There are some currencies that have a ticker symbol with just two letters ([Aeternity](https://etherscan.io/token/0x5ca9a71b1d01849c0a95490cc00559717fcf0d1d) for example).

sangaman replied:
I was not aware, and there are actually a handful that have 4 letters as well, so I'll relax this to be a length of 2-4.

michael1011 commented on lib/service/Service.ts:
```diff
@@ -55,9 +55,13 @@ const argChecks = {
+    }
+  },
+  VALID_NETWORK_CLIENT: ({ networkClient }: { networkClient: number }) => {
+    if (NetworkClients[networkClient] === undefined) throw errors.INVALID_ARGUMENT('network client is not recognized');
```
What about just `if (!NetworkClients[networkClient])`

michael1011 commented on lib/utils/utils.ts:
```diff
@@ -134,3 +135,14 @@ export const groupBy = (arr: object[], keyGetter: (item: any) => string | number
   return Date.now();
 };
+
+export const getEnumNumbers = (enumObj: any) => {
```
Is this function used somewhere?

sangaman replied:
Oops nope, I was using it at one point for putting restrictions on the eligible network clients in the database, but abandoned that approach. Thanks.

michael1011 commented on lib/cli/commands/addcurrency.ts:
```diff
@@ -0,0 +1,38 @@
+    description: 'The contract address for layered tokens such as ERC20',
+    type: 'string',
+  },
+  subunits: {
```
I am fine with `subunits` but did you consider using the decimals of the currency instead?

sangaman replied:
What do you mean? Like the base 10 exponent for the number of subunits in a unit? Such that bitcoin's would be 8? 10^8 satoshis in a bitcoin.

michael1011 replied:
Yeah exactly. Especially because all of those ERC20 contracts are using that format and copying `100000000000000000` is harder than just `18`.

sangaman replied:
Yeah I agree that's better, I'll change it.

michael1011 commented on lib/cli/commands/addcurrency.ts:
```diff
@@ -0,0 +1,38 @@
+  network_client: {
+    description: 'The payment channel network client',
+    type: 'string',
+    choices: ['LND', 'RAIDEN'],
```
Yeah makes sense. We should add `choices` to `logLevel` too.

michael1011 commented on lib/service/Service.ts:
```diff
@@ -51,9 +52,13 @@ const argChecks = {
-    const supportedCurrencies = ['BTC', 'LTC']; // TODO: detect supported currencies automatically instead of hardcoding
-    if (!supportedCurrencies.includes(currency)) throw errors.INVALID_ARGUMENT(`currency ${currency} is not supported`);
+  VALID_CURRENCY: ({ currency }: { currency: string }) => {
+    if (currency.length < 2 || currency.length > 4 || !currency.match(/^[A-Z]+$/))  {
```
I have to annoy you again but I just noticed that there are tokens with symbols that consist of 5 characters: `STEEM`, `WAVES`, `STRAT`

What about getting rid of the upper limit entirely?

sangaman commented:
OK, I changed `subunits` to `decimalPlaces` and I also renamed `NetworkClients` to `SwapClients` as we'd discussed in gitter. See what you think @michael1011, thanks.

michael1011 commented on lib/cli/commands/addcurrency.ts:
```diff
@@ -0,0 +1,38 @@
+
+export const command = 'addcurrency <currency> <swap_client> [decimal_places] [token_address]';
+
+export const describe = 'add a supported currency';
```
`add a supported currency` sounds like you add an already supported currency. What about `add support for a currency` or simply `add a currency`?

sangaman replied:
I'll also change the comments in the proto definition.

michael1011 commented on lib/cli/commands/addcurrency.ts:
```diff
@@ -0,0 +1,38 @@
+  },
+  decimal_places: {
+    description: 'The places to the right of the decimal point of the smallest subunit (e.g. satoshi)',
+    default: 8,
```
Idea for the future: automatically detect `decimal_places` of ERC20s if nothing is specified. The users are running an Ethereum node anyways.

michael1011 commented on lib/orderbook/OrderBook.ts:
```diff
@@ -85,14 +92,21 @@ class OrderBook extends EventEmitter {
-      this.pairIds.push(pair.id);
+    currencies.forEach((currency) => {
+      this.currencies.set(currency.id, currency);
+    });
```
`currencies.forEach(currency => this.currencies.set(currency.id, currency));`?

michael1011 commented on lib/orderbook/OrderBook.ts:
```diff
@@ -135,6 +149,56 @@ class OrderBook extends EventEmitter {
+    if (currency) {
+      for (const pairId of this.pairIds) {
+        if (pairId.includes(currencyId)) {
+          throw errors.CURRENCY_CANNOT_BE_REMOVED(currencyId, pairId);
```
If the currency is in multiple pairs the user has to call `removeCurrency` and `removePair` multiple times to know the blocking pairs and remove them. I think all affected pairs should be returned at once. But that can be done in a separate PR.

Btw: what is more efficient `pairId.includes(currencyId)` or checking if `currencyId` equals either `baseCurrency` or `quoteCurrency`?

sangaman replied:
I was curious and it looks like the latter is indeed faster: http://jsben.ch/nIxYb

I'm sure it won't make a difference but I'll have it use the faster way.

michael1011 commented on lib/orderbook/OrderBook.ts:
```diff
@@ -154,7 +218,7 @@ class OrderBook extends EventEmitter {
 
     if (orderId === undefined) {
-      throw errors.INVALID_PAIR_ID(pairId);
+      throw errors.UNSUPPORTED_PAIR(pairId);
```
Just because the `orderId` can't be found that doesn't mean that the pair is not supported?

sangaman replied:
Looks like this was using the wrong error message before this PR, but I'll fix it.

michael1011 commented on lib/service/Service.ts:
```diff
@@ -86,6 +91,25 @@ class Service extends EventEmitter {
+    await this.orderBook.addCurrency({
+      tokenAddress,
+      swapClient,
+      decimalPlaces: decimalPlaces || 8,
```
I think this default value should be in the orderbook instead of the service class.

sangaman commented:
I changed a bit more of the wording around "supported" currencies/pairs in comments and ran through manual tests of all the calls once more. I think this is ready to merge.

sangaman commented:
The build isn't starting because [travis is having issues](https://www.traviscistatus.com/) but it passes all tests locally.
